src/sdk/service/network-validation-service/network-validation-service.ts

```diff
@@ -0,0 +1,333 @@
+    const mask = parseInt('80000000', 16) >> (cidr.hostIdentifier - 1);
+    const netmask = new Netmask(this.getIntegerAsIPv4Address(mask));
+    return {
+      gateway: cidr.address,
+      netmask: netmask.address
+    };
+  }
+
+  /**
+   * Gets an IP range object from a string of the format: a.b.c.d - e.f.g.h
+   * @param {string} rangeStr: the range string to parse
+   * @return {IPv4AddressRange} range object of the form stated in the description or null if invalid.
+   */
+  public static getIPv4AddressRangeFromString(rangeStr: string): IPv4AddressRange | null {
+    const components = rangeStr.split('-');
+    if (components.length === 2) {
+      const start = new IPv4Address(components[0].trim()); // Validate and get address string.
+      const end = new IPv4Address(components[1].trim());
+      const startInt = this.getIPv4AsInteger(start);
+      const startPos = startInt >= 0;
+      const endInt = this.getIPv4AsInteger(end);
+      const endPos = endInt >= 0;
+      // test whether start IP is actually lower than the end IP.
+      if ((startPos === endPos && endInt >= startInt) ||
+        (startPos !== endPos && startInt < endInt)) {
+        return {
+          start: start.address,
+          end: end.address
+        };
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Takes a netmask and any IP address on the network and converts it to a CIDR
+   * string format. If either the netmask or IP address are invalid, returns
+   * null.
+   * @param {string} netmask: the netmask string.
+   * @param {string} ipAddress: the IP address string.
+   * @return {string} a CIDR formatted network representation string.
+   */
+  public static getSubnetCidr(netmask: Netmask, ipAddress: IPv4Address): string | null {
+    const prefixSize = this.getNetmaskPrefixSize(netmask);
+    const components = ipAddress.address.split('.');
+    const prefixMod = Math.floor((prefixSize - 1) / 8);
+    let subnetIp = '';
+    for (let i = 0; i < 4; i++) {
+      const val = components[i];
+      if (i < Math.ceil(prefixMod)) {
+        subnetIp += components[i];
+      } else {
+        subnetIp += parseInt(val, 10) & (-1 << (8 * (i + 1)) - prefixSize);
+      }
+      if (i !== 3) {
+        subnetIp += '.';
+      }
+    }
+    return subnetIp + '/' + prefixSize;
+  }
+
+  /**
+   * Gets the range of a subnet from the netmask and any IP address on the
+   * network. If either the netmask or IP address are invalid, returns null.
+   * @param {string} netmask: the netmask string.
+   * @param {string} ipAddress: the IP address string.
+   * @return {IPv4AddressRange} an object of the form { start: "x.x.x.x", end: "x.x.x.x" },
+   *         where the start field contains the IP address at the low end of the
+   *         range and the end field contains the IP address at the high end of
+   *         the range.
+   */
+  public static getSubnetIPv4AddressRange(netmask: Netmask, ipAddress: IPv4Address): IPv4AddressRange | null {
```

Comment:
type union with null for the return type is no longer valid. Check the docs and return types of all methods